### PR TITLE
Solve cache not writable on docker

### PR DESCRIPTION
**Pull request by Stoakes**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #192
| License       | MIT
| Doc PR        |  x

On docker run, start.sh launch a composer install as root user. Therefore, apache user can't read var/cache/dev content.

To fix it, I get apache user and apache group in `/etc/apache2/apache2.conf` and chown var folder with those values.

### CODE REVIEW

**Comment by soyuka** [comment]
Not sure that this should be fixed from the `start.sh` file, see [this file](https://github.com/Stoakes/api-platform/blob/163969f52bb29fd72e0724c6666473ec0978694c/docker/apache/start_safe_perms)

File: docker/start.sh [hunk 1]
```diff
@@ -7,9 +7,16 @@ export DOCKER_BRIDGE_IP=$(ip ro | grep default | cut -d' ' -f 3)
 if [ "$SYMFONY_ENV" = 'dev' ]; then
     composer install --prefer-dist --no-progress --no-suggest
 else
-    composer install --prefer-dist --no-dev --no-progress --no-suggest --optimize-autoloader --classmap-authoritative
+    composer install --prefer-dist --no-dev --no-progress --no-suggest --optimize-autoloader
```
Comment (Simperfit) [inline docker/start.sh]: why did you remove `--classmap-authoritative` ?

File: docker/start.sh [hunk 2]
```diff
@@ -7,9 +7,16 @@ export DOCKER_BRIDGE_IP=$(ip ro | grep default | cut -d' ' -f 3)
...
 fi
 
+#get apache user : second word of the first line (from docker/apache/start_safe_perms)
```
Comment (Simperfit) [inline docker/start.sh]: It does not fix the issue when we are using `docker-compose exec web composer update`

File: docker/start.sh (hunk 1 above)
Comment (Stoakes) [inline docker/start.sh]: My mistake,
Copy/paste error.

Comment (theofidry) [inline docker/start.sh]: Indeed it shouldn't be change, `--optimize-autoloader` can be removed though as `--classmap authoritative` does imply `--optimize-autoloader`.

Comment (Stoakes) [inline docker/start.sh]: Solved.

File: docker/start.sh (hunk 2 above)
Comment (Stoakes) [inline docker/start.sh]: Sure.

But except with --user flag, I think there is no way to run a command as a non-root user.

**Comment by dunglas** [comment]
Wouldn't be better to run the composer install with the apache user instead of doing this?

File: docker/start.sh [hunk 3]
```diff
@@ -10,6 +10,13 @@ else
     composer install --prefer-dist --no-dev --no-progress --no-suggest --optimize-autoloader --classmap-authoritative
 fi
 
+#get apache user : second word of the first line (from docker/apache/start_safe_perms)
+apache_user=$(head -n 1 /etc/apache2/apache2.conf| cut -d " " -f 2)
```
Comment (dunglas) [inline docker/start.sh]: It cannot work because the apache user and group is changed after that: https://github.com/Stoakes/api-platform/blob/163969f52bb29fd72e0724c6666473ec0978694c/docker/apache/start_safe_perms#L26-L32

**Comment by soyuka** [comment]
> Wouldn't be better to run the composer install with the apache user instead of doing this?
 
+1 everything should be done as one user, ie `www-data`.

**Comment by Stoakes** [comment]
Running everything as www-data would be easier.

In fact I don't know why api-platform does all those stuff around owner and group, I was just trying to fix the issue.
 As @soyuka pointed out, it would be easier (and less hacky) to manage permissions in start_safe_perms.

So what's next ? Do I update this PR or close it in favour of a whole docker system rewrite ?

**Comment by dunglas** [comment]
@Stoakes because we use volumes for dev, and we cannot change the owner of files mounted as a volume.

**Comment by soyuka** [comment]
Hmm, I don't get why this is needed.

What I do in my own config:

```
version: '2'
services:
  php:
    container_name: api_platform_php
    build: php-fpm
    ports:
      - "9000:9000"
    volumes:
      - ./symfony:/var/www/symfony
  nginx:
    container_name: api_platform_nginx
    build: nginx
    ports:
      - "8081:80"
    depends_on:
      - php
    volumes:
      - ./symfony:/var/www/symfony
      - ./logs/nginx/:/var/log/nginx
```

Where `./symfony` is a symlink to the project:

```
lrwxrwxrwx 1 abluchet abluchet   21 Aug  1 17:40 symfony -> /home/abluchet/project
```

Then, everything runs as `www-data`, I have no hacks whatsoever. The only thing I do in both Dockerfiles is to set the same `uid`:

```
RUN usermod -u 1000 www-data
```

I could post my docker configuration if you want.

**Comment by dunglas** [comment]
@soyuka are you using Linux or Mac?

**Comment by soyuka** [comment]
Both linux @ work and os x @ home. This works on both.

**Comment by dunglas** [comment]
Yes but with our trick we doesn't need to set the uid.

**Comment by soyuka** [comment]
You mean using a 50 lines bash script instead of 1 line in the Dockerfile? ^^ Anyway it's different here because I just need nginx and php-fpm to run on the same user. Though I don't need anything for my volume permission.

**Comment by dunglas** [comment]
@soyuka it's very important for the first impression to have something working out of the box, without requiring any weird config, even for people not confortable with Docker.

**Comment by soyuka** [comment]
> @soyuka it's very important for the first impression to have something working out of the box, without requiring any weird config, even for people not confortable with Docker.

100% agreed

**Comment by soyuka** [comment]
Okay, so I've spent a few hours playing with your docker configuration. I see a few issues but those can not be solved properly with the current  `FROM php:7.0-apache`.

The hack in `start_safe_perm` is not necessary (oh and btw it comes from [here](https://github.com/kojiromike/docker-magento/blob/master/apache/start_safe_perms)) if we use the `envvars` file that is sourced by `apache2-foreground`, a binary declared by https://github.com/docker-library/php. 

We could make this work if we were not installing the composer dependencies within docker. Indeed, by doing so, if we want correct permissions, we need to issue the `composer` command from `www-data`. If it's run by `root` directories will have the wrong permissions. For example, as `composer` post install scripts are creating the cache directories, it'll fail.

So what about running everything as `www-data` as I suggested?

 Because, the docker source image we're using is exposing the `80` port, which is restricted to root. Though, it would just work great with the `8080` port which would be forwarded to `80` by docker-compose (host has root capabilities). The thing is, it's hard coded and there are no ways to change it (check the related issues on https://github.com/docker-library/php). 

What now?

A cleaner hack is to use a simple `su -s /bin/bash -c "composer command" www-data` when we need to so that the permissions are maintained.

IMHO, it would be best to make this work by using `php-fpm` and `nginx`, because they're not using any unsafe port. Or, we just use another `php` image that doesn't require the `80` port ([example with alpine + nginx](https://github.com/xataz/dockerfiles/blob/master/nginx-php/Dockerfile)).

I'm adding a pull-request now.

**Comment by dunglas** [comment]
@soyuka as we're in the process of merging #162 who's already using FPM, can you push commits (or comment) in this branch instead?

**Comment by soyuka** [comment]
Okay I'll take a look at it!

**Comment by Simperfit** [comment]
Since #162 has been merge, it is still relevant ?
